### DISCUSSION TITLE
JReleaser process Killed: 9 on macOS

### DISCUSSION THREAD

**Discussion by aalmiray** [Homebrew/discussions#2794] [Tap maintenance and Homebrew development]
Hello everyone,

I've got a problem with a formula hosted at a custom tap (https://github.com/jreleaser/homebrew-tap). This formula unpacks a Java distribution (launcher script + JAR files) that also bundle a custom Java Runtime created with jlink. In this way jreleaser may be used without a previous Java runtime installed.

I've verified that the Java distribution contained by the Zip file works as expected. I've also manually unpacked the Zip file downloaded by brew and confirmed that it works. However using the installed package results in `Killed: 9`.

I've looked for answers and most answers say reinstalling xcode tools and/or brew make it work, sadly that's not the case. I've not found other formulae where a custom Java Runtime is included.

The full formula definition is:

```ruby
# Generated with JReleaser 0.10.0-SNAPSHOT at 2021-12-28T14:52:05.57129Z
class Jreleaser < Formula
  desc "Release projects quickly and easily with JReleaser"
  homepage "https://jreleaser.org"
  version "0.10.0"
  license "Apache-2.0"

  if OS.linux? && Hardware::CPU.arm? && Hardware::CPU.is_64_bit?
    url "https://github.com/jreleaser/jreleaser/releases/download/v0.10.0/jreleaser-standalone-0.10.0-linux-aarch64.zip"
    sha256 "3518365f5f093bb6947b170fcc96da87589d5d589550d7d0fd64d3deeaa17bbd"
  end
  if OS.linux? && Hardware::CPU.intel?
    url "https://github.com/jreleaser/jreleaser/releases/download/v0.10.0/jreleaser-standalone-0.10.0-linux-x86_64.zip"
    sha256 "51efafbc49dc035d2c89ae2ada9745c975f7a3e8fb54f1c608325d230ba908dc"
  end
  if OS.mac? && Hardware::CPU.arm?
    url "https://github.com/jreleaser/jreleaser/releases/download/v0.10.0/jreleaser-standalone-0.10.0-osx-aarch64.zip"
    sha256 "1881fc7a88fd0ea687c9ff96aaca17821744ab5fb0813a85462578f990419c37"
  end
  if OS.mac? && Hardware::CPU.intel?
    url "https://github.com/jreleaser/jreleaser/releases/download/v0.10.0/jreleaser-standalone-0.10.0-osx-x86_64.zip"
    sha256 "184540ecf756d688babbed32ef520042290fad21b5a38266d15143cc5d101439"
  end


  def install
    libexec.install Dir["*"]
    bin.install_symlink "#{libexec}/bin/jreleaser"
  end

  test do
    output = shell_output("#{bin}/jreleaser --version")
    assert_match "0.10.0", output
  end
end
```

Additional context -> https://github.com/jreleaser/jreleaser/issues/675

**Comment by carlocab** [comment]
Does `Killed: 9` occur on both ARM and Intel?

**Reply by aalmiray** [reply]
I can only tell for Intel as I don't have access to ARM.

**Reply by carlocab** [reply]
I see. It's hard to say what's happening. I'd try to see if using the runtime from one of the `openjdk` formulae works, and/or trying to attach the relevant process to a debugger to see if I can get a backtrace.

**Reply by aalmiray** [reply]
You'd have to strip the custom Java Runtime from the unpacked archive. Previously JReleaser used to ship just the JARs with no bundled Java Runtime, it also had a dependency on the openjdk formula. Then we switched to bundling the Java Runtime so that you won't need an extra formula installed, and this is when the problem started.

**Reply by carlocab** [reply]
> You'd have to strip the custom Java Runtime from the unpacked archive.

Yes, you would have to do that. There are a handful of formulae in Homebrew/core we do something like that with.

**Comment by aalmiray** [comment]
As an additional data point, every single `*.dylib` provided by the custom Java Runtime bundled with this formula is altered by the install procedure performed by `libexec.install Dir["*"]` and that's probably the cause of the problem.

If so, how could these files be skipped form any alterations during install?

**Reply by SMillerDev** [reply]
Software should not be providing it's own Java runtime, we'll need a way to strip that out or stop shipping the software in homebrew.

**Reply by aalmiray** [reply]
Why is that the case? The whole reason to have `jlink` as a standard tool since Java 9 is to provide a custom Java runtime that ships with your program to avoid consumers having to install a matching Java runtime before hand. jlink also produces a reduced runtime image that includes only what you need to run the program instead of a full blown JRE/JDK.

**Reply by SMillerDev** [reply]
Because homebrew already ships a Java runtime and that is the one that should be used.

https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae

**Reply by aalmiray** [reply]
Yes, I'm aware that Homebrew already ships a Java runtime but that's not enough.

That Java runtime is fixed to a particular vendor (AdoptOpenJDK/Adoptoum/Temurin). What If I need a different JR provided by a another vendor?
That Java runtime is a full fledged one. What If I need a smaller one with additional custom Java modules (security, etc)?

Reusing the provided JR by Homebrew restricts those options.

**Reply by SMillerDev** [reply]
If it can't run with the dependencies homebrew provides it either shouldn't be in homebrew, or those dependencies should be adjusted to make them more broadly usable.

Vendoring dependencies with all compatibility and security issues of the vendored version is not the preferred solution in homebrew.

**Reply by aalmiray** [reply]
Perhaps, if the formula were to be provided by core where strict rules must be followed. But's that's not the case here as I'm using a tap.

**Reply by carlocab** [reply]
> That Java runtime is fixed to a particular vendor (AdoptOpenJDK/Adoptoum/Temurin). What If I need a different JR provided by a another vendor?
> 
> That Java runtime is a full fledged one. What If I need a smaller one with additional custom Java modules (security, etc)?
> 
> 
> 
> Reusing the provided JR by Homebrew restricts those options.

Why isn't your bundled runtime subject to the same limitations?

**Reply by aalmiray** [reply]
Because as the developer of the app and JR bundle I can determine if the given bundle meets the expectations or not. Not 100% fool proof of course (nothing is).
